nevergrad/optimization/optimizerlib.py

```diff
@@ -1522,35 +1531,63 @@ def _internal_provide_recommendation(self) -> ArrayLike:
+            mutated_sigma = self.sigma * np.exp(self._rng.normal(0, 1) / np.sqrt(self.dimension))
+            individual = self.current_center + mutated_sigma * self._rng.normal(0, 1, self.dimension)
+            tmp = mutated_sigma
+        else:
+            individual = self.current_center + self.sigma * self._rng.normal(0, 1, self.dimension)
         parent = self.parents[self.num_ask % len(self.parents)]
         candidate = parent.spawn_child().set_standardized_data(individual, reference=self.parametrization)
         if parent is self.parametrization:
             candidate.heritage["lineage"] = candidate.uid
-        candidate._meta["sigma"] = self.sigma
+        candidate._meta["sigma"] = tmp
         return candidate
 
     def _internal_tell_candidate(self, candidate: p.Parameter, value: float) -> None:
         candidate._meta["loss"] = value
+        if self.populationSizeAdapatation:
+            self.popsize.add_value(value)
         self.children.append(candidate)
-        if len(self.children) >= self.llambda:
+        popSize = -1
+        mu = -1
+        if self.populationSizeAdapatation:
+            popSize = self.popsize.llambda
+            mu = self.popsize.mu
+        else:
+            popSize = self.llambda
+            mu = self.mu
+        if len(self.children) >= popSize:
             # Sorting the population.
             self.children.sort(key=lambda c: c._meta["loss"])
             # Computing the new parent.
-            self.parents = self.children[: self.mu]
+            self.parents = self.children[: mu]
             self.children = []
             self.current_center = sum(c.get_standardized_data(reference=self.parametrization)  # type: ignore
-                                      for c in self.parents) / self.mu
-            # EMNA update
-            stdd = [(self.parents[i].get_standardized_data(reference=self.parametrization) - self.current_center)**2
-                    for i in range(self.mu)]
-            if self.isotropic:
-                self.sigma = np.sqrt(sum(stdd) / (self.mu * self.dimension))
+                                      for c in self.parents) / mu
+            if self.populationSizeAdapatation:
+                if self.popsize.llambda < self.paraLimit * self.dimension: # Population size not large enough for emna
+                    if self.isotropic:
+                        self.sigma = np.exp(np.sum(np.log([c._meta["sigma"] for c in self.parents])) / mu)
+                    else:
+                        self.sigma = np.exp(np.sum(np.log([c._meta["sigma"] for c in self.parents]), axis=0) / mu)
+                else:
+                    stdd = [(self.parents[i].get_standardized_data(reference=self.parametrization) - self.current_center)**2 for i in range(mu)]
+                    if self.isotropic:
+                        self.sigma = np.sqrt(sum(stdd) / (mu * self.dimension))
```

Comment:
same as above, and to share as much code as possible then maybe just use a `axis` variable:
`axis = 0 if self.isotropic else None`